### PR TITLE
refactor(blocksync): break up `poolRoutine`

### PR DESCRIPTION
**Pull request by melekes**
Replaces https://github.com/cometbft/cometbft/pull/2152

Co-authored by: @faddat 

---

#### PR checklist

- [ ] ~~Tests written/updated~~
- [ ] ~~Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)~~
- [ ] ~~Updated relevant documentation (`docs/` or `spec/`) and code comments~~
- [x] Title follows the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) spec

### CODE REVIEW

**Comment by melekes** [comment]
🤔 the e2e tests are failing for some reason

**Comment by faddat** [comment]
Hey thank you for the fixes @melekes !

File: internal/blocksync/reactor.go [hunk 1]
```diff
@@ -561,3 +427,174 @@ func (bcR *Reactor) BroadcastStatusRequest() {
...
+		ChannelID: BlocksyncChannel,
+		Message:   &bcproto.BlockRequest{Height: request.Height},
+	})
+	if !queued {
+		bcR.Logger.Debug("Send queue is full, drop block request", "peer", peer.ID(), "height", request.Height)
+	}
+}
+
+func (bcR *Reactor) handleBlockRequestsRoutine() {
+	go func() {
```
Comment (cason) [inline internal/blocksync/reactor.go:446]: We could remove the `go func()` here and just call `go r.handleBlockRequestsRoutine()` on the original loop.

File: internal/blocksync/reactor.go [hunk 2]
```diff
@@ -561,3 +427,174 @@ func (bcR *Reactor) BroadcastStatusRequest() {
...
+	if missingExtension {
+		bcR.Logger.Info(
+			"no extended commit yet",
+			"last_block_height", state.LastBlockHeight,
+			"vote_extensions_disabled", voteExtensionsDisabled,
+			"blocks_synced", blocksSynced,
+		)
+		return true
+	}
+	return false
```
Comment (cason) [inline internal/blocksync/reactor.go:505]: We always return `missingExtension`, don't we?

File: internal/blocksync/reactor.go [hunk 3]
```diff
@@ -473,68 +395,12 @@ FOR_LOOP:
...
-				if state.ConsensusParams.Feature.VoteExtensionsEnabled(first.Height) {
-					err = extCommit.EnsureExtensions(true)
-				} else if extCommit != nil {
-					err = fmt.Errorf("received non-nil extCommit for height %d (extensions disabled)", first.Height)
-				}
-			}
-			if err != nil {
+
+			if state, err = bcR.processBlock(first, second, firstParts, state, extCommit); err != nil {
 				bcR.Logger.Error("Error in validation", "err", err)
```
Comment (cason) [inline internal/blocksync/reactor.go:400]: #2872 has changed this log message, improving it. Recall to keep the new message.

**Review by cason — APPROVED** [review]
Thanks for this.

I could reproduce the changes in the original code, and the behavior matches.

**Pull request merged**